### PR TITLE
k8s logtail

### LINKED ISSUE #3429: [k8s] Log SkyPilot output to container stdout
SkyPilot pods are run with `sleep infinity` to let SkyPilot runtime manage the lifecycle of the pod. As a result, container logs are empty `kubectl get logs`.

This causes problems when integrating with other log management systems running on the cluster, which directly scrape from the container logs. 

From user(s):
> We are using FileBeat for logs collection. Basically, all the pods' logs are automatically wrote into the dir /var/log/containers/ by k8s (the dir is on the host where the pod is running on), and we just configure this dir into FileBeat.
I found that when we use Sky, all the log files of LLM instances under /var/log/containers/ are just empty because Sky redirects logs
> Another user stated they use external logging system to collect logs of pods

Can we replace the `sleep infinity` command with something which monitors `~/sky_logs` and runs `tail` on any files that get created. Will need to handle multiple tasks executing simultaneously since their logs may get interleaved. 

## Workaround for now
Mount the `~/sky_logs`  to a local hostPath on the host machine. For example, if you add this to your `~/.sky/config.yaml`:
```
kubernetes:
  pod_config:
    spec:
      containers:
        - volumeMounts:
            - mountPath: /home/sky/sky_logs
              name: sky-logs-volume
              readOnly: false
      volumes:
        - name: sky-logs-volume
          hostPath:
            path: /var/log/sky_logs
            type: Directory
```

it will write all logs to /var/log/sky_logs on the host machine. Then you can add /var/log/sky_logs as a scrape path on your logging system. This is similar to how k8s writes all logs to /var/log/containers/.

For example, here's the directory structure that gets created when running two sky clusters independently:
```
/var/log/sky_logs$ tree # This is on the host machine
.
├── sky-2024-04-04-12-32-13-597156
│   ├── run.log
│   └── tasks
│       └── run.log
└── sky-2024-04-04-12-33-24-369770
    ├── run.log
    └── tasks
        └── run.log
```
Make sure you create the dir and make sure it has appropriate permissions (`mkdir -p /var/log/sky_logs; chmod -R 777 /var/log/sky_logs`) before mounting it.

### PR DESCRIPTION
**Pull request by asaiacai**
Closes #3429

This swaps the container sleep with a logtailing process so that pod logs now show task logs. These are retrievable via `kubectl logs <POD_NAME>` and also written to `/var/log/pods/*/*/*.log` on the node disk, so it should integrate with most k8s logging solutions out of the box (OpenTelemetry, Fluentd, Filebeats, etc.)

Implemented as a watcher process that does `tail -f` if there's an open writing file handle on the task log by checking `lsof` and closes the `tail -f` program if there are no such writers.

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] All smoke tests: `pytest tests/test_smoke.py` 
- [x] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_container_logs_multinode_kubernetes` 
- [x] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_container_logs_two_jobs_kubernetes` 
- [x] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_container_logs_two_simultaneous_jobs_kubernetes`

### CODE REVIEW

**Comment by asaiacai** [comment]
@romilbhardwaj I finished running the smoke tests on my end if you have a spare cycle to take a look

**Review by romilbhardwaj — COMMENTED** [review]
Thanks @asaiacai!

File: sky/templates/kubernetes-ray.yml.j2 [hunk 1]
```diff
@@ -316,8 +316,48 @@ available_node_types:
           image: {{image_id}}
           # Do not change this command - it keeps the pod alive until it is
           # explicitly killed.
-          command: ["/bin/bash", "-c", "--"]
-          args: ['trap : TERM INT; sleep infinity & wait;']
+          command: ["/bin/bash", "-c"]
+          args: 
+            - |
+              FILE_PATTERN="~/sky_logs/*/tasks/*.log"
```
[the remark this answers is not in the document]
Comment (asaiacai) [reply sky/templates/kubernetes-ray.yml.j2]: sounds good. i added this back.

[the remark this answers is not in the document]
Comment (asaiacai) [reply sky/templates/kubernetes-ray.yml.j2]: done. wrapped the logtail in a function with `|| sleep infinity` fallback.

**Comment by asaiacai** [comment]
@romilbhardwaj reran the smoke tests and manually checked `kubectl logs` again but let me know and I can draft up some additional tests.

**Comment by romilbhardwaj** [comment]
Thanks @asaiacai. Pushed some changes to reduce sleep (which was causing delays in pod termination). lmk what you think. 

BTW, tests seem to be failing for me:
```
$ pytest tests/test_smoke.py::test_container_logs_two_simultaneous_jobs_kubernetes --kubernetes
```
Logs: https://gist.github.com/romilbhardwaj/4065c409f089760fb11415f0bf186262

Looks like logs are correctly tailed on the first `sky launch`, but logs from any subsequent `sky exec` are not tailed.

This was on a GKE cluster. Interestingly, the manual repro seems to work on my local `sky local up` cluster, but the test still fails at `kubectl get pods  | grep t-container-logs-two-5n-e5 |  grep head | awk '{print $1}' | xargs -I {} kubectl logs {} | wc -l | grep 18`.

**Review by romilbhardwaj — APPROVED** [review]
Thanks @asaiacai! Tested it, works nicely. LGTM.

**Pull request merged**